### ISSUE TITLE
`go test - race` free nbdserver 

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-Disk#239]
components:
- [x] lba
- [x] deduped storage
- [x] non deduped storage (https://github.com/zero-os/0-Disk/issues/239#issuecomment-306119463)
- [x] tlog storage, in memory sequence cache, and data history
- [x] in memory storage (https://github.com/zero-os/0-Disk/issues/239#issuecomment-306119955)

while working on #232, i found that `go test -race` on nbdserver directory reveals many data races.
full logs : https://gist.github.com/iwanbk/d1726cc202ec60757f3d8a0fe94ee740

some snippets

```
WARNING: DATA RACE
Write at 0x00c420872458 by goroutine 153:
  github.com/zero-os/0-Disk/nbdserver/lba.(*LBA).Set()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/lba/lba.go:88 +0x187
  github.com/zero-os/0-Disk/nbdserver/ardb.(*dedupedStorage).Set()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/deduped.go:54 +0x1e2
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorageDeadlock.func1()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:450 +0xbd

Previous write at 0x00c420872458 by goroutine 214:
  github.com/zero-os/0-Disk/nbdserver/lba.(*LBA).Set()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/lba/lba.go:88 +0x187
  github.com/zero-os/0-Disk/nbdserver/ardb.(*dedupedStorage).Set()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/deduped.go:54 +0x1e2
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorageDeadlock.func1()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:450 +0xbd

Goroutine 153 (running) created at:
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorageDeadlock()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:466 +0x2c6
  github.com/zero-os/0-Disk/nbdserver/ardb.TestDedupedDeadlock()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/deduped_test.go:175 +0x125
  testing.tRunner()
      /home/ibk/go/src/testing/testing.go:657 +0x107

Goroutine 214 (running) created at:
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorageDeadlock()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:466 +0x2c6
  github.com/zero-os/0-Disk/nbdserver/ardb.TestDedupedDeadlock()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/deduped_test.go:175 +0x125
  testing.tRunner()
      /home/ibk/go/src/testing/testing.go:657 +0x107
==================
```

```
WARNING: DATA RACE
Read at 0x00c420a2d5c8 by goroutine 187:
  github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).Close()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:220 +0x80
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorage()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:206 +0x1261
  github.com/zero-os/0-Disk/nbdserver/ardb.testTlogStorage()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog_test.go:212 +0x242
  github.com/zero-os/0-Disk/nbdserver/ardb.TestTlogStorageWithInMemory()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog_test.go:27 +0x1dc
  testing.tRunner()
      /home/ibk/go/src/testing/testing.go:657 +0x107

Previous write at 0x00c420a2d5c8 by goroutine 1007:
  github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).GoBackground()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:254 +0x204

Goroutine 187 (running) created at:
  testing.(*T).Run()
      /home/ibk/go/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /home/ibk/go/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /home/ibk/go/src/testing/testing.go:657 +0x107
  testing.runTests()
      /home/ibk/go/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /home/ibk/go/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/zero-os/0-Disk/nbdserver/ardb/_test/_testmain.go:90 +0x20f

Goroutine 1007 (running) created at:
  github.com/zero-os/0-Disk/nbdserver/ardb.testBackendStorage()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/ardb_test.go:76 +0xe6
  github.com/zero-os/0-Disk/nbdserver/ardb.testTlogStorage()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog_test.go:212 +0x242
  github.com/zero-os/0-Disk/nbdserver/ardb.TestTlogStorageWithInMemory()
      /home/ibk/pathgo/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog_test.go:27 +0x1dc
  testing.tRunner()
      /home/ibk/go/src/testing/testing.go:657 +0x107
==================
```

**Comment by iwanbk** [comment]
Didn't find data race for this test
```
go test -race -run ^TestNondeduped
```

**Comment by iwanbk** [comment]
after passed the `go test -race` in our development machine, we might need to enable it at travis.
(and makes it much more slower )

**Comment by iwanbk** [comment]
`go test -race -run ^TestInMemory` doesn't failed on master branch

**Comment by GlenDC** [comment]
lba data race fixed by #242

**Comment by GlenDC** [comment]
+ deduped is race-free (according to `go test -race`), as that seems to be caused by the `lba data race`, which was fixed since #242;
+ tlogStorage is race-free since #240;

**Comment by GlenDC** [comment]
Fixed by #240 and #242.
